### PR TITLE
UAR-1354 sign out validation and back link fix fir remove journey

### PR DESCRIPTION
**Pull request by markpit**
### JIRA link
https://companieshouse.atlassian.net/browse/UAR-1354

### Change description
- Fix to make the sign-out 'save' button appear for remove journey
- Fix the sign-out link to make it work for remove journey urls that have more slashes eg, .../remove/somepage
- Use the Registration approach to get the previous page from the request headers and use that for the back link / No button on the sign-out screen

### Work checklist

- [ ] Tests added where applicable
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: views/sign-out.html [hunk 1]
```diff
@@ -21,7 +21,7 @@
 
 {% if saveAndResume and (journey == "register") %}
   {% set hintText = "We will save your application." %}
-{% elif saveAndResume and (journey == "update") %}
+{% elif saveAndResume and (journey == "update" or journey == "remove") %}
```
Comment (mwestacott) [inline views/sign-out.html:24]: Is it maybe worth defining an `isUpdateOrRemoveJourney` variable at the top of this file to shorten this repeating logic check in the four times it's done? Or could you just do `journey != "register"` ? Of course, logic will further simplify if/when the non 'save and resume' code is ever ripped out...

File: views/includes/sign-out-user-banner.html [hunk 2]
```diff
@@ -1,8 +1,17 @@
 {% if userEmail %}
+
+  {% set signOutPath = 'sign-out?page=' + templateName %}
+
+  {% if journey === 'remove' %}
+    {% set url = url if url else OE_CONFIGS.UPDATE_AN_OVERSEAS_ENTITY_URL %}
```
Comment (mwestacott) [inline views/includes/sign-out-user-banner.html:6]: This just looks a bit confusing - setting 'update' as the url if we're on a remove journey. I assume this is handling reuse of the update pages in the remove journey, but worth a comment I think, explaining this for future readers of this code (or me, in a week's time when I've forgotten most things apart from my name).

File: src/controllers/update/update.sign.out.controller.ts [hunk 3]
```diff
@@ -2,15 +2,24 @@ import { NextFunction, Request, Response } from "express";
...
+import { getPreviousPageUrl, isRemoveJourney } from "../../utils/url";
 
 export const get = (req: Request, res: Response, next: NextFunction) => {
+  let journey = config.JourneyType.update;
+  let previousPage = `${config.UPDATE_AN_OVERSEAS_ENTITY_URL}${req.query["page"]}`;
+
+  if (isRemoveJourney(req)) {
+    journey = config.JourneyType.remove;
+    previousPage = getPreviousPageUrl(req, config.UPDATE_AN_OVERSEAS_ENTITY_URL);
+  }
```
Comment (dlloyd1-ch) [inline src/controllers/update/update.sign.out.controller.ts:14]: Might be worth a comment explaining why a remove journey needs to call the getPreviousPageUrl function whereas the update journey does not. Or maybe a comment on the function itself.

Comment (markpit) [reply src/controllers/update/update.sign.out.controller.ts:16]: added a comment to say why we use that approach

**Review by dlloyd1-ch — APPROVED** [review]

**Pull request merged**